templates/install_puppet.ps1.erb

```diff
@@ -0,0 +1,185 @@
+.Parameter UseLockedFilesWorkaround
+  Set to $true to enable execution of the puppetres.dll move workaround. See https://tickets.puppetlabs.com/browse/MODULES-4207
+#>
+[CmdletBinding()]
+param(
+  # PuppetPID _must_ come first!, this script needs PuppetPID to be a positional parameter to execute
+  # correctly from the module.
+  [String] $PuppetPID,
+  [String] $Source='<%= @_msi_location %>',
+  [String] $Logfile='<%= @_logfile %>',
+  [String] $InstallDir='<%= @install_dir %>',
+  [String] $PuppetMaster='<%= @_puppet_master %>',
+  [String] $PuppetStartType='<%= @_agent_startup_mode %>',
+  [String] $InstallArgs='<%= @_install_options.map do |option|
+                                if option.class == Hash
+                                  key_value = option.shift
+                                  "#{key_value[0]}=\"#{key_value[1]}\""
+                                else
+                                  option
+                                end
+                              end.join(' ') %>',
+  [String] $InstallScriptPIDFile='<%= @_install_pid_file_loc %>',
+  [Bool]   $UseLockedFilesWorkaround=<%= @_move_dll_workaround %>
+)
+# fetch_install_dir, move_puppetres_dll and restore_puppetres_dll serve as a workaround for older
+# installations of puppet: we used to need to move puppetres.dll out of the way during puppet
+# upgrades because the file would lock and cause network stack restarts.
+# See https://tickets.puppetlabs.com/browse/MODULES-4207
+<#
+.Synopsis
+  Fetch the location of the puppet installation from the registry
+#>
+function Script:fetch_install_dir {
+  param()
+  begin {
+    if (Get-ItemProperty -Path "HKLM:\SOFTWARE\Puppet Labs\Puppet" -Name RememberedInstallDir64 -ErrorAction SilentlyContinue) {
+      return (Get-ItemProperty -Path "HKLM:\SOFTWARE\Puppet Labs\Puppet").RememberedInstallDir64
+    } elseif (Get-ItemProperty -Path "HKLM:\SOFTWARE\Puppet Labs\Puppet" -Name RememberedInstallDir -ErrorAction SilentlyContinue) {
+      return (Get-ItemProperty -Path "HKLM:\SOFTWARE\Puppet Labs\Puppet").RememberedInstallDir
+    } else {
+      return $null
+    }
+  }
+}
+
+<#
+.Synopsis
+  Move/rename puppetres.dll to a temporary location
+#>
+function Script:move_puppetres_dll {
+  param()
+  begin {
+    $rand_string = [String[]](Get-Random)
+    $temp_puppetres = "$env:temp\puppetres$rand_string.dll"
+    # _Never_ use the $InstallDir top-level parameter to try and find the InstallDir
+    # for the puppetres workaround. The workaround should _always_ fetch the InstallDir
+    # from the registry. This is so a user can specify a different InstallDir than the
+    # directory where the current package is installed and the workaround will still work
+    # for the already installed package
+    $InstallDir = fetch_install_dir
+    if (Test-Path "$InstallDir\puppet\bin\puppetres.dll") {
+      Out-File -InputObject "Moving puppetres.dll to $temp_puppetres" -FilePath $Logfile -Append

```

Comment:
The canonical way of doing this is normally;

`"Moving puppetres.dll to $temp_puppetres"  | Out-File -FilePath $Logfile -Append"`

Reply:
May be easier to abstract this into a method? `Log <error_message>` is more readable.

Reply:
+1 to refactor to it's own function